pkg/controller/stacks/persona/persona.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persona
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/stacks"
+)
+
+const (
+	managedRolesLabel   = "rbac.crossplane.io/managed-roles"
+	managedRolesEnabled = "true"
+
+	reconcileTimeout = 1 * time.Minute
+
+	loggerName = "stacks/namespace-personas"
+
+	adminPersona = "admin"
+	editPersona  = "edit"
+	viewPersona  = "view"
+
+	errFailedToCreateClusterRole  = "failed to create clusterrole"
+	errFailedToDeleteClusterRoles = "failed to delete clusterroles"
+	errFailedToGetNamespace       = "failed to get namespace"
+
+	logFailedToCreateDuringSync = "failed to create during sync"
+	logFailedToDeleteDuringSync = "failed to delete during sync"
+)
+
+var (
+	personas = []string{adminPersona, editPersona, viewPersona}
+
+	resultRequeue = reconcile.Result{Requeue: true}
+)
+
+// Reconciler reconciles Namespaces
+type Reconciler struct {
+	kube client.Client
+	log  logging.Logger
+	factory
+}
+
+// Setup adds a controller that reconciles Namespaces.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &nsPersonaHandlerFactory{},
```

Comment:
is a factory strictly necessary in this reconciler? i've used that pattern before when the "handler" can only be generated at runtime with specific config or environment information, but the handler in this reconciler looks like it just takes in some objects and does no other init work that you'd want to avoid during tests (e.g. making a cloud provider call, or looking up something in the k8s api).

you could consider removing this factory and just use a handler directly (instead of delegating the creation of the handler to the factory) to reduce complexity. I believe that would work. Your choice though, it's not a big deal.

Reply:
I took a quick pass at removing the factory but I ended up with a very similar complexity as I still needed a way to inject a handler into the tests.  If you are ok with leaving the handler in I think we can revisit this in future refactors.

Reply:
yes sure, if it isn't an obvious win then no problem leaving it in!